### PR TITLE
Adds the unwieldy element and an admin only sledgehammer 

### PR DESCRIPTION
**Pull request by Ryll-Ryll**
## About The Pull Request
I originally did this with my wounds PR because I wanted to add a sledgehammer as a non-sharp counterpart to the fireaxe for breaking bones, but I figure I may as well atomize. This adds an unwieldy element, mainly for 2h weapons, that crudely emulates swinging a heavy weapon when attacking mobs. How crudely, you ask? This crudely!

[](https://i.imgur.com/alF2qxr.png)

Depending if the weapon is in your left or right hand, the element will check the tile counter-clockwise or clockwise from you to your target for any solid obstructions. If there's something solid there, your attack hits the obstruction instead, and you're briefly stunned standing for half a second. Otherwise, the hit goes through and you switch hands on the weapon, so as to swing from the other side next. Fun!

## Why It's Good For The Game
Seems like it might be a fun mechanic to throw on big dangerous weapons like sledgehammers or giant swords/whatever to compensate for their raw killing/crippling power by making them more dangerous to use in cramped areas. I'm interested in hearing feedback on whether this might be a fun thing to throw on large, dangerous weapons maybe in exchange for some other buffs, or if people think it might be useful for other new weapons/concepts.

### CODE REVIEW

**Comment by necromanceranne** [comment]
Is there no way for an overhead smash? this seems like it'll get pretty weird with how maint is laid out. Maybe if you aim for the head?

**Comment by PKPenguin321** [comment]
I dont like that it stuns you on a whiff, if it instead greatly raised your cooldown before you could attack again that would be better

**Comment by actioninja** [comment]
maybe make you drop it or something like what pkp suggested, this just makes it ridiculously risky to actually use anything like this.

**Comment by Ryll-Ryll** [comment]
It's not a paralyze "you fall over and are helpless" stun, it's the "you can't move + drop what you're holding" stun for half a second, just to clarify. You can also define your own behavior for each weapon, this is just what I came up with on the spot for the sledgehammer. I can easily change it to change cooldown or just make you drop it OR immobilize you for half a second, or whatever else would  be fitting.

**Comment by TemporalOroboros** [comment]
So goliath tendril stun?

**Comment by necromanceranne** [comment]
The concept of it stopping an attack based on what is around you is cool but making it a stun is probably not the best. Cooldown seems more fitting. And a loud _bonk_ sound to go with it.

**Comment by Fikou** [comment]
maybe make it so that if you aim for the head after a 1-2 second do after you swing from above?

**Comment by Qustinnus** [comment]
Ironically this would make it extremely hard to smash windows with this hammer even though it's purpose is to demolish things. I'd still like to see a solution for that tbf

**Comment by Ryll-Ryll** [comment]
@Qustinnus nah i made it only check when you're swinging at a mob because i realized that issue t he first time I hit a window with it.

**Comment by Qustinnus** [comment]
based

**Comment by PKPenguin321** [comment]
Really though, this shouldn't apply to attacking downed mobs or structures because think about it, you can just overhead swing

File: code/datums/elements/unwieldy.dm [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+/datum/element/unwieldy/Attach(datum/target)
+	. = ..()
+	if(!isitem(target))
+		return ELEMENT_INCOMPATIBLE
+
+	RegisterSignal(target, COMSIG_ITEM_PRE_ATTACK, .proc/swing)
+
+/datum/element/unwieldy/Detach(datum/target)
+	UnregisterSignal(target, list(COMSIG_ITEM_PRE_ATTACK))
```
Comment (Qustinnus) [inline code/datums/elements/unwieldy.dm:21]: list isn't needed

File: code/datums/elements/unwieldy.dm [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	else
+		var/highest_layer = 0
+		for(var/i in tile_check)
+			var/atom/thing = i
+			if(thing.density && thing.layer > highest_layer)
+				obstacle = thing
+				highest_layer = thing.layer
+
+	if(obstacle)
+		SEND_SIGNAL(I, COMSIG_UNWIELDY_BONK, user, intended_target, obstacle)
```
Comment (Qustinnus) [inline code/datums/elements/unwieldy.dm:52]: In cases like this you want a callback, as all you're doing is running behavior on the object the component is on.

File: code/game/objects/items/sledgehammer.dm [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	wielded = TRUE
+	AddElement(/datum/element/unwieldy)
+
+/// triggered on unwield of two handed item
+/obj/item/sledge/proc/on_unwield(obj/item/source, mob/user)
+	wielded = FALSE
+	RemoveElement(/datum/element/unwieldy)
+
+/obj/item/sledge/update_icon_state()
+	icon_state = "hammeroff"
```
Comment (Qustinnus) [inline code/game/objects/items/sledgehammer.dm:43]: whats the point in this?

File: code/game/objects/items/sledgehammer.dm [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	force = 5
+	throwforce = 15
+	w_class = WEIGHT_CLASS_BULKY
+	slot_flags = ITEM_SLOT_BACK
+	attack_verb = list("battered", "crushed", "slammed", "hammered")
+	hitsound = 'sound/weapons/smash.ogg'
+	max_integrity = 200
+	armor = list("melee" = 0, "bullet" = 0, "laser" = 0, "energy" = 0, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 100, "acid" = 30)
+	resistance_flags = FIRE_PROOF
+	var/wielded = FALSE // track wielded status on item
```
Comment (Qustinnus) [inline code/game/objects/items/sledgehammer.dm:20]: don't do this, it's poor use of components as you shouldn't need to save the state of the component.

File: code/game/objects/items/sledgehammer.dm [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+/obj/item/sledge/proc/bonk(obj/item/I, mob/living/user, atom/intended_target, atom/obstruction)
+	user.visible_message("<span class='danger'>[user] slams [src] into [obstruction] mid-swing!</span>", "<span class='danger'><b>Your swing intended for [intended_target] is cut short by [obstruction]!</b></span>")
+	playsound(get_turf(src), 'sound/effects/clang.ogg', 100, TRUE, -1)
+	obstruction.attackby(src, user)
+	user.Stun(0.5 SECONDS)
+	if(isturf(obstruction))
+		user.do_attack_animation(obstruction)
+
+/obj/item/sledge/afterattack(atom/A, mob/living/carbon/user, proximity)
```
Comment (Qustinnus) [inline code/game/objects/items/sledgehammer.dm:53]: this proc seems like it should be part of the unwieldy behavior, or be a callback thats run on the unwieldy component after a succesful hit.

File: code/modules/mob/inventory.dm [hunk 6]
```diff
@@ -169,6 +169,13 @@
 		return hand_index || TRUE
 	return FALSE
 
+/mob/proc/force_hand_swap()
```
Comment (Qustinnus) [inline code/modules/mob/inventory.dm:172]: This skips over the proper procs for equipin and unequiping things from slots, and I don't think this is the right way of handling this.

Comment (Qustinnus) [reply code/modules/mob/inventory.dm:172]: haha get it, handling? ok ill stop.

**Review by ghost — CHANGES_REQUESTED** [review]
Getting there but you need to use callbacks, hmu if you need help with that.

File: code/game/objects/items/sledgehammer.dm (hunk 5 above)
Comment (Qustinnus) [reply code/game/objects/items/sledgehammer.dm:53]: On that note, that means you can also include things like the direction etc in the callback which means you dont have to do the calculations again.

**Comment by Ryll-Ryll** [comment]
@Qustinnus thank you for the thorough reviews! I'm gonna mothball this PR for now since I don't have any immediate plans for stuff using this and there's lots of other PR's I have open already, but rest assured when I come back to this I'll thank you again for the review!

**Pull request closed without merging**